src/background_workers.c

```diff
@@ -245,130 +385,148 @@ void *RedisAI_Run_ThreadMain(void *arg) {
+          queueItem *next_item = queuePop(run_queue_info->run_queue);
+          RedisAI_RunInfo *next_rinfo = (RedisAI_RunInfo *)next_item->value;
+          // Push the DAG to the front of the queue, and then the item we just
+          // popped in front of it, so that it becomes the first item in the queue.
+          // The rationale is, since the DAG needs to wait for other workers, we are
+          // giving way to the next item and we'll get back to the DAG when that is done
+          queuePushFront(run_queue_info->run_queue, evicted_rinfo);
+          queuePushFront(run_queue_info->run_queue, next_rinfo);
+        }
+        // If there's nothing else in the queue
         else {
-          // If the client is still connected
-          if (evicted_rinfo->client) {
-            // Unblock the client. Note that we do this for every evicted info in a batch,
-            // which will be associated with a different client. Here we unblock every
-            // client appropriately
-            RedisModule_UnblockClient(evicted_rinfo->client, evicted_rinfo);
+          // We push the DAG back at the front
+          queuePushFront(run_queue_info->run_queue, evicted_rinfo);
+          // Since there's nothing else on the queue we just break out and give other
+          // workers a chance to produce the inputs needed for this DAG step
+          array_free(evicted_items);
+          break;
+        }
+      }
+
+      // If the op was ran successfully and without any error, then put the entry back
+      // on the queue unless all ops for the device have been executed
+      if (do_run == 1 && run_error == 0) {
+        // Here we iterate backwards to keep the first evicted on top
+        // A side effect of this is that we are potentially changing priority in the queue
+        // We could solve this using a priority queue, TODO for later
+        for (long long i = array_len(evicted_items) - 1; i >= 0; i--) {
+          // Get the current evicted run info
+          RedisAI_RunInfo *evicted_rinfo = (RedisAI_RunInfo *)(evicted_items[i]->value);
+          // If the DAG for the top-most item is complete for the device, then
+          // we don't push it back on the queue
+          if (i ==0 && device_complete_after_run == 1) {
+            continue;
           }
+          queuePushFront(run_queue_info->run_queue, evicted_rinfo);
+        }
+      }
+
+      // TODO now we can figure out of the device is complete or the dag is complete
+      // if (dag_complete_op_count == evicted_rinfo[0]->dagOpCount) -> ublock, free
+      // if (dag_device_complete_op_count == evicted_rinfo[0]->dagDeviceOpCount) -> device complete
+
+      // If there's nothing else to do for the DAG in the current worker or if an error
+      // occurred in any worker, we just move on
+      if (device_complete == 1 || device_complete_after_run || do_unblock == 1 || run_error == 1) {
+        for (long long i=0; i<array_len(evicted_items); i++) {
```

Comment:
Do all these conditions suggest that there was a run? I understand correctly? so `if(do_run) {free each evicted item}`

Reply:
Getting back to the previous comment, a rinfo will be placed back in the queue until all ops have been executed for the device. This condition reads: if all ops for the device have been executed, or if the request needs to be unblocked, or there was an error, it's time to deallocate the rinfo.